keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,823 @@
+    returns (GetGroupSnapshotResponse) {
+        option (alpha_method) = true;
+    }
+  …
+}
+```
+
+#### CreateVolumeGroup
+
+This RPC will be called by the CO to create a new volume group on behalf of a user.
+This operation MUST be idempotent. If a volume corresponding to the specified volume name already exists, is compatible with the specified parameters in the CreateVolumeGroupRequest, the Plugin MUST reply 0 OK with the corresponding CreateVolumeGroupResponse.
+CSI Plugins MAY create the following types of volume groups:
+
+* Create a new empty volume group.
+* Create a new volume group from a list of existing volumes.
+* At restore time, create a single volume from individual snapshot and then join an existing group.
+ * Create an empty group
+ * Create a volume from snapshot in the group
+
+The following are non-goals:
+* Non goal: Create a new group and at the same time create a list of new volumes in the group.
+* Non goal: Create a group from an existing group snapshot in one step.
+
+```
+message CreateVolumeGroupRequest {
+  option (alpha_message) = true;
+
+  // suggested name for volume group (required for idempotency)
+  // This field is REQUIRED.
+  string name = 1;
+
+  // params passed from VolumeGroupClass
+  // This field is OPTIONAL.
+  map<string, string> parameters = 2;
+
+  // volume_ids from existing volumes.
+  // Empty volume group will be created if volume_ids is not specified.
+  // This field is OPTIONAL.
+  repeated string volume_ids = 3;
+}
+
+message CreateVolumeGroupResponse {
+  option (alpha_message) = true;
+
+  // Contains all attributes of the newly created volume group.
+  // This field is REQUIRED.
+  VolumeGroup volume_group = 1;
+}
+
+message VolumeGroup {
+  option (alpha_message) = true;
+
+  // The identifier for this volume group, generated by the plugin.
+  // This field is REQUIRED.
+  string volume_group_id = 1;
+
+  // Opaque static properties of the volume group.
+  // This field is OPTIONAL.
+  map<string, string> volume_group_context = 2;
+
+  // Underlying volumes in this group. The same definition in CSI Volume.
+  // This field is OPTIONAL.
+  repeated .csi.v1.Volume volumes = 3;
+}
+```
+
+#### CreateVolume
+
+1. When a new volume is created with a volume group id parameter, the volume will be created and added to the existing volume group.
+2. A new volume can also be created without a volume group id parameter. It can be added to a volume group later through the ModifyVolumeGroup RPC.
+
+Note that for filesystems based storage systems, only option 1 can be supported. For block based storage systems. Both option 1 and 2 may be supported. However there is a possibility that option 2 will not work for ConsistencyGroups as the volume is created without the consideration of which group the volume will be placed in.
+
+```
+message CreateVolumeRequest {
+      string name = 1;
+      …
+      string volume_group_id = 8 [(alpha_field) = true];
```

Comment:
If one Volume could join multiple VolumeGroups, this field type should be repeated string?

Reply:
Yes.